net: DHCPv4: option callback

m-byte opened the pull request:
I would like to propose an option callback API for the DHCPv4 client. The callback provides a clean interface to register custom options in the client and return the values of those to the application. For example, this can be used to assign the SNTP server IP from the DHCP server.

The sample and test have been updated.

The code has been tested with qemu_x86 and the attached test.

Please let me know any questions, feedbacks or concerns.

rlubos commented on include/zephyr/net/dhcpv4.h:
```diff
@@ -44,6 +98,44 @@ enum net_dhcpv4_state {
 /** @endcond */
 
+/**
+ * @brief Helper to initialize a struct gpio_callback properly
```
It's `struct net_dhcpv4_option_callback`

m-byte replied:
Sorry, will fix it

rlubos commented on subsys/net/ip/dhcpv4.c:
```diff
@@ -880,14 +886,48 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
-		default:
-			NET_DBG("option unknown: %d", type);
+		default: {
+			struct net_dhcpv4_option_callback *cb, *tmp;
```
The commit message says:
> Callbacks can only be registered for options that are not handled
internally (see `dhcpv4_parse_options` in dhcpv4.c for more info).

Which is a consequence of processing callbacks in the `default` label (unrecognized option). I wonder, is it really needed to introduce such a limitation? Unless I miss something, it should be possible to register callback for any option, if the option processing was moved before the "internal" processing (`switch` instruction). The net_pkt cursor is backed up, so it should be fairly transparent.

rlubos replied:
Additional remark - i think such a functionality wouldn't be used that often, so it'd make sense to add a Kconfig option for it, just so that we don't increase the library footprint when not needed.

m-byte replied:
Thanks for your review! I really like your suggestion of enabling this for any option. Even though it wouldn't be necessary for my use case, this might still be useful to others. I'll refactor it so that this is moved before the `switch` block.

rlubos commented on subsys/net/ip/dhcpv4.c:
```diff
@@ -1236,6 +1276,30 @@ static void dhcpv4_start_internal(struct net_if *iface, bool first_start)
+	__ASSERT(cb, "No callback!");
+	__ASSERT(cb->handler, "No callback handler!");
+
+	sys_slist_prepend(&option_callbacks, &cb->node);
```
`sys_slist` is not thread safe, so any operations on it should be protected. I think `lock` mutex defined in the file could be used here (it's already locked for instance when we loop over the callbacks above).

m-byte replied:
Ok, will be fixed

rlubos commented on subsys/net/ip/dhcpv4.c:
```diff
@@ -1236,6 +1276,30 @@ static void dhcpv4_start_internal(struct net_if *iface, bool first_start)
+	__ASSERT(cb, "No callback!");
+	__ASSERT(cb->handler, "No callback handler!");
+
+	if (!sys_slist_is_empty(&option_callbacks)) {
```
`sys_slist_is_empty()`  seems redundant here, you could just call `sys_slist_find_and_remove()` directly - if the list is empty, it'll fail anyway.

m-byte replied:
As you noticed above, I started with the gpio_callback as reference. That's where this originates from. I'll simplify it.

rlubos commented on samples/net/dhcpv4_client/src/main.c:
```diff
@@ -57,6 +61,16 @@ static void handler(struct net_mgmt_event_callback *cb,
 	}
 }
 
+static void option_handler(const uint8_t option,
+	     struct net_dhcpv4_option_callback *cb,
+	     size_t length)
```
nit: Please align to the opening bracket.

m-byte replied:
Sorry about that, I forgot about it in the sample

rlubos commented on samples/net/dhcpv4_client/src/main.c:
```diff
@@ -69,6 +83,11 @@ int main(void)
 	iface = net_if_get_default();
 
+	net_dhcpv4_init_option_callback(&dhcp_cb, option_handler,
+					42, ntp_server, sizeof(ntp_server));
```
Please define a symbol for the magic number (42).

m-byte replied:
Will do

rlubos commented on include/zephyr/net/dhcpv4.h:
```diff
@@ -25,6 +25,60 @@ extern "C" {
+ * CONTAINER_OF() if original struct net_dhcpv4_option_callback is stored in
+ * another private structure.
+ */
+typedef void (*net_dhcpv4_option_callback_handler_t)(const uint8_t option,
```
Isn't `option` redundant here? It's already part of the `struct net_dhcpv4_option_callback`

rlubos replied:
IMO it'd make sense to include `struct net_if *` pointer instead, in case there are multiple network interfaces in the system.

m-byte replied:
Good points, I will address them

rlubos commented on subsys/net/ip/dhcpv4.c:
```diff
@@ -880,14 +886,48 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
+						return false;
+					}
+
+					cb->handler(type, cb, MIN(cb->max_length, length));
```
Another concern I have - currenly the option cb will be called for any message type. This could cause abiguity, in case there are more DHCP servers in the network (as theoretically we can receive mutliple OFFER messages). Perhaps the callback shoudl only be called on ACK messages?

m-byte replied:
Wouldn't that concern apply to the way e.g. subnet mask and router are currently assigned? From what I can tell, they also don't have any extra checks in place.
Either way, it would be possible to pass `msg_type` to the callback.

m-byte replied:
Please let me know if you believe the updated code sufficiently addresses this

rlubos replied:
That is actually a very good point and something that may have been overlooked - probably because the ACK message would typically be the last one, but temporarily we could have got incorrect configuration stored in the system. 

As for the callback, I think that passing message type to the cb is also fine.

m-byte committed:
net: DHCPv4: Add option callback

In certain scenarios, it may be necessary to get values of additional
options from the application layer. With this patch, this can be
accomplished by registering a callback with the DHCP client.

This change has been tested using the posix build in qemu.

Signed-off-by: Matthias Breithaupt <m.breithaupt@vogl-electronic.com>

m-byte committed:
samples: dhcpv4_client: Add option callback

Extend the DHCPv4 sample to make use of the option callback functionality.
This serves as a reference of how to register and use a custom option
callback.

Signed-off-by: Matthias Breithaupt <m.breithaupt@vogl-electronic.com>

m-byte committed:
test: DHCPv4: Add test for option callbacks

Add tests for the option callback API. The test uses a previously unused
option already contained in the sample offer.

Signed-off-by: Matthias Breithaupt <m.breithaupt@vogl-electronic.com>

rlubos commented on include/zephyr/net/dhcpv4.h:
```diff
@@ -44,6 +44,121 @@ enum net_dhcpv4_state {
+ */
+typedef void (*net_dhcpv4_option_callback_handler_t)(struct net_dhcpv4_option_callback *cb,
+						     size_t length,
+						     enum net_dhcpv4_msg_type *msg_type,
```
I think it'd be better to provide the value, not the pointer. We don't want the app to mess with its content.

rlubos commented on subsys/net/ip/dhcpv4.c:
```diff
@@ -1236,6 +1273,35 @@ static void dhcpv4_start_internal(struct net_if *iface, bool first_start)
+
+int net_dhcpv4_add_option_callback(struct net_dhcpv4_option_callback *cb)
+{
+	__ASSERT(cb, "No callback!");
+	__ASSERT(cb->handler, "No callback handler!");
```
It's a public API, so it'd be better if the function returned an error instead of asserting.

rlubos commented on subsys/net/ip/dhcpv4.c:
```diff
@@ -1236,6 +1273,35 @@ static void dhcpv4_start_internal(struct net_if *iface, bool first_start)
+
+int net_dhcpv4_remove_option_callback(struct net_dhcpv4_option_callback *cb)
+{
+	__ASSERT(cb, "No callback!");
```
Same here.

rlubos commented on subsys/net/ip/dhcpv4.c:
```diff
@@ -881,13 +896,18 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
 
 			if (net_pkt_skip(pkt, length)) {
-				NET_DBG("option unknown, skip err");
+				if (unhandled) {
+					NET_DBG("option unknown, skip err");
+				} else {
+					NET_DBG("option callback, skip err");
+				}
```
IMO this If here is not needed, the log is just reporting `net_pkt_skip()` error.

rlubos commented on subsys/net/ip/dhcpv4.c:
```diff
@@ -881,13 +896,18 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
 		default:
-			NET_DBG("option unknown: %d", type);
+			if (unhandled) {
+				NET_DBG("option unknown: %d", type);
```
I'd rather add `else` here, with `option unknown, handled by callback` or similar.

m-byte replied:
Ok, I'll add the else above and remove the whole if statement here.

rlubos commented on subsys/net/ip/dhcpv4.c:
```diff
@@ -707,6 +700,28 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
 
+
+#if defined(CONFIG_NET_DHCPV4_OPTION_CALLBACKS)
+		net_pkt_cursor_backup(pkt, &backup);
```
You should re-set `unhandled = true` for each loop iteration (each new option).

rlubos reviewed (commented):
Thank you for addressing my initial feedback, left a few more comments.

jukkar commented on include/zephyr/net/dhcpv4.h:
```diff
@@ -44,6 +44,121 @@ enum net_dhcpv4_state {
+ * Note: To help setting it, see net_dhcpv4_init_option_callback() below
+ */
+struct net_dhcpv4_option_callback {
+	/** This is meant to be used internally and the user should not
+	 * mess with it.
+	 */
+	sys_snode_t node;
+
+	/** Actual callback function being called when relevant. */
+	net_dhcpv4_option_callback_handler_t handler;
+
+	/** The DHCP option this callback is attached to. */
+	uint8_t option;
+
+	/** Maximum length of data buffer. */
+	size_t max_length;
+
+	/** Pointer to a buffer of size max_length that is used to store the
+	 * option data.
+	 */
+	void *data;
```
To me it looks like the intention was to keep this as "internal" data that the user is not suppose to touch. Perhaps we could place the content to the private header next to the source files to enforce this.

m-byte replied:
If this were to be placed in a private header, things like removing a specific callback will be more complicated. Also, all functions that currently accept the struct will have to be changed since the application won't know about the struct.
The current implementation is based on how things are done in the GPIO callback API.

jukkar commented on include/zephyr/net/dhcpv4.h:
```diff
@@ -44,6 +44,121 @@ enum net_dhcpv4_state {
+ * @param max_length The maximum length of the data returned.
+ */
+static inline void net_dhcpv4_init_option_callback(struct net_dhcpv4_option_callback *callback,
+						   net_dhcpv4_option_callback_handler_t handler,
```
If the callback struct is "public" then the handler parameter is kind of unnecessary as it is part of callback anyway. But I would suggest the keep this prototype like this but make `struct net_dhcpv4_option_callback` private instead.

m-byte replied:
Since this function is used to populate only the fields that should be changed by the user application, the whole point is to call it with a callback pointer and all the data that will be placed inside.
This is how things are also handled for the GPIO callbacks: https://github.com/zephyrproject-rtos/zephyr/blob/d01780fc9403166d250ded3d02372a7e1a9db4fd/include/zephyr/drivers/gpio.h#L1440

m-byte replied:
Doesn't work, since the user code then doesn't know the storage size of this struct. And I can't think of any non-hacky way around that.

jukkar replied:
Fair enough, I was just wondering whether we need to fully expose the struct internals. At least we should mark the callback struct internal as `/** @cond INTERNAL_HIDDEN */ .... /** @endcond */` so that the documentation generator does not try to document the internal fields.

m-byte replied:
I'm fine with that and will update it.

jukkar reviewed (approved):
LGTM
